### ISSUE TITLE
[spike] Using Fleet to automate OS upgrades

### ISSUE THREAD

**Issue by kkaempf** [rancher-sandbox/cOS-toolkit#339]
GitOps with [Fleet](https://fleet.rancher.io) can already be used to configure and upgrade the application layer.

Goal of this research spike is to utilize Fleet to trigger OS upgrades.

Apparently Darren is already working in this direction. Get in contact with him first !

### Acceptance criteria

- [ ] reach out to Darren ( https://github.com/ibuildthecloud/os2 )
- [ ] demo showing Fleet used to trigger an OS upgrade
- [ ] documentation about the demo setup
- [ ] demo should partly work on baremetals (TBD)
- [ ] (if needed) creation of further cOS issues to address cOS deficits

**Comment by mudler** [comment]
Darren is unavailable for few days, I'll start to have a look for documenting a Fleet+k3s deployment on top of cOS

**Comment by mudler** [comment]
Note, to build https://github.com/ibuildthecloud/os2 from my PR ( https://github.com/ibuildthecloud/os2/pull/3 ):

```
export REPO=<IMAGE>
export LABEL=<TAG>
make build
docker push $REPO:$LABEL
make iso
```

**Comment by mudler** [comment]
So I've exec'd into the rancher pod and created the admin user with `ensure-default-admin` and password can be reset with `reset-password`. although I'm still seeing sporadic failures in the webui and I can't create/import new cluster due to permission issues.  I think I should have seen the cluster being imported automatically instead, and the rancherd logs are still full of the logs above, investigating.

**Comment by mudler** [comment]
Unrelated note to the above, os2 seems to ship a custom version of https://github.com/rancher/system-upgrade-controller which should take care of the upgrade part to be triggered with fleet.

**Comment by davidcassany** [comment]
>  I'll start to have a look for documenting a Fleet+k3s deployment on top of cOS, following more or less https://github.com/rancher-sandbox/cOS-toolkit/wiki/Github-runner-on-k8s#cos-configuration style

@mudler I had to adapt several part of this yaml to deploy k3s, are you still using this documented yaml? I managed to deploy k3s and manually install fleet on top. I am trying to understand what fleet is how it could hypothetically be used to manage system upgrades. At glance it looks like this is out of scope of Fleet.

**Comment by mudler** [comment]
That's what I've used to deploy k3s+fleet:

```yaml
name: "Fleet master"
stages:
   initramfs:
     - name: "Setting"
       hostname: "rancher-fleet-master"
   network:
     - commands:
       - cp -rf /etc/sysconfig/network/ifcfg-eth0 /etc/sysconfig/network/ifcfg-eth1
       - systemctl restart network
     - name: "Setup k3s"
       if: '[ -z "$(blkid -L COS_SYSTEM || true)" ]'
       directories:
       - path: "/usr/local/bin"
         permissions: 0755
         owner: 0
         group: 0
       commands:
       - |
            curl -sfL https://get.k3s.io | \
            INSTALL_K3S_VERSION="v1.20.4+k3s1" \
            INSTALL_K3S_EXEC="--tls-san {{.Values.node.hostname}}" \
            INSTALL_K3S_SELINUX_WARN="true" \
            sh -
       - curl https://raw.githubusercontent.com/helm/helm/master/scripts/get-helm-3 | bash
       - |
          helm -n fleet-system install --create-namespace --wait \
              fleet-crd https://github.com/rancher/fleet/releases/download/v0.3.3/fleet-crd-0.3.3.tgz
          helm -n fleet-system install --create-namespace --wait \
              fleet https://github.com/rancher/fleet/releases/download/v0.3.3/fleet-0.3.3.tgz
     - name: "Setup SSH keys"
       authorized_keys:
         rancher:
         - github:mudler
```

I'm trying to run os2 instead, afaics it have all the bits to make upgrade with fleet work. 

1) There is the system-upgrade-controller which accepts `Plans` which are system upgrades. In our case should be a plan calling

**Comment by davidcassany** [comment]
> That's what I've used to deploy k3s+fleet:
> 
> ```yaml
> name: "Fleet master"
> stages:
>    initramfs:
>      - name: "Setting"
>        hostname: "rancher-fleet-master"
>    network:
>      - commands:
>        - cp -rf /etc/sysconfig/network/ifcfg-eth0 /etc/sysconfig/network/ifcfg-eth1
>        - systemctl restart network
>      - name: "Setup k3s"
>        if: '[ -z "$(blkid -L COS_SYSTEM || true)" ]'
>        directories:
>        - path: "/usr/local/bin"
>          permissions: 0755
>          owner: 0
>          group: 0
>        commands:
>        - |
>             curl -sfL https://get.k3s.io | \
>             INSTALL_K3S_VERSION="v1.20.4+k3s1" \
>             INSTALL_K3S_EXEC="--tls-san {{.Values.node.hostname}}" \
>             INSTALL_K3S_SELINUX_WARN="true" \
>             sh -
>        - curl https://raw.githubusercontent.com/helm/helm/master/scripts/get-helm-3 | bash
>        - |
>           helm -n fleet-system install --create-namespace --wait \
>               fleet-crd https://github.com/rancher/fleet/releases/download/v0.3.3/fleet-crd-0.3.3.tgz
>           helm -n fleet-system install --create-namespace --wait \
>               fleet https://github.com/rancher/fleet/releases/download/v0.3.3/fleet-0.3.3.tgz
>      - name: "Setup SSH keys"
>        authorized_keys:
>          rancher:
>          - github:mudler
> ```
> 
> I'm trying to run os2 instead, from my digging it does have all the bits to make upgrade with fleet work.
> 
>     1. There is the [system-upgrade-controller](https://github.com/rancher/system-upgrade-controller) which accepts `Plans` which are system upgrades. In our case should be a plan calling something that upgrades from a container which is the new system, e.g. https://github.com/ibuildthecloud/os2/blob/master/files/usr/sbin/suc-upgrade
> 
>     2. The fleet role here is the ability to propagate the Plan to all the downstream clusters
> 
> 
> That's what I could guess from the code and the bits around. My problem is I can't get fleet/rancher correctly up after deploying it (I'm running a VM in openstack).
> 
> For the openstack setup, I've uploaded the latest master qcow image of cOS, and after booting it up, I run `cos-upgrade --docker-image --no-verify quay.io/costoolkit/test-images:os2`, at that point after a reboot rancherOS should start.
> I'm using this as userdata when creating the VM:
> 
> ```yaml
> name: "Rancherd master (control plane - up)"
> 
> # Uncomment above if loading from userdata
> role: "cluster-init"
> 
> stages:
>    initramfs:
>      - name: "Setting"
>        hostname: "rancher-fleet-master"
>      - name: "Setup distro"
>        systemd_firstboot:
>         keymap: us
>         locale: en_US.UTF-8
>         timezone: UTC
> 
>    network:
>      - commands:
>        # DHCP for eth1 as well
>        - cp -rf /etc/sysconfig/network/ifcfg-eth0 /etc/sysconfig/network/ifcfg-eth1
>        - systemctl restart network
>        systemctl:
>          start:
>           - systemd-timesyncd
> ```

Oh, I see I was missing that part...

**Comment by mudler** [comment]
Ok, I've managed to run an upgrade without Fleet:

1) Create a VM on openstack from the qcow images in the artifacts. Use the following cloud-init to deploy the instance:

```yaml
name: "Fleet master"
stages:
   initramfs:
     - name: "Setting"
       if: '[ -z "$(blkid -L COS_SYSTEM || true)" ]'
       hostname: "rancher-fleet-master"
       commands:
       - |
          target=/usr/local/.ros-state

          # Always want the latest update of systemd conf from the image
          mkdir -p ${target}/etc/systemd/
          rsync -av /etc/systemd/ ${target}/etc/systemd/

          # Only populate ssh conf once
          if [ ! -e ${target}/etc/ssh ]; then
            mkdir -p ${target}/etc/ssh/
            rsync -a /etc/ssh/ ${target}/etc/ssh/
          fi

          sed -i '/overlay \/home /d' /etc/fstab
          sed -i '/overlay \/opt /d' /etc/fstab
          nsenter -m -t 1 -- umount /sysroot/home
          nsenter -m -t 1 -- umount /sysroot/opt

          # setup directories as persistent
          for i in root opt home var/lib/rancher var/lib/kubelet etc/systemd etc/rancher etc/ssh usr/libexec var/log var/lib/wicked; do
            mkdir -p ${target}/$i /$i
            nsenter -m -t 1 -- mount /sysroot${target}/$i /sysroot/$i -t none -o bind
          done

          # This is hidden so that if you run some selinux label checking or relabeling the bind
          # mount won't screw up things.  If you have two files at different paths they will get
          # labeled with two different labels.
          mkdir -p ${target}/empty
          nsenter -m -t 1 -- mount /sysroot${target}/empty /sysroot${target} -o bind,ro

          # ensure /var/log/journal exists so it's labeled correctly
          nsenter -m -t 1 -- mkdir -p /sysroot/var/log/journal
   initramfs.after:
   - if: '[ -z "$(blkid -L COS_SYSTEM || true)" ]'
     commands:
     - restorecon -R -v /etc /home /opt /var /usr/local /tmp /srv /root
   fs.before:
   - if: '[ -z "$(blkid -L COS_SYSTEM || true)" ]'
     commands:
     - restorecon -R -v /etc /home /opt /var /usr/local /tmp /srv /root

   network:
     - commands:
       - cp -rf /etc/sysconfig/network/ifcfg-eth0 /etc/sysconfig/network/ifcfg-eth1
       - systemctl restart network
     - name: "Setup k3s"
       if: '[ -z "$(blkid -L COS_SYSTEM || true)" ]'
       directories:
       - path: "/usr/local/bin"
         permissions: 0755
         owner: 0
         group: 0
       commands:
       - |
            curl -sfL https://get.k3s.io | \
            INSTALL_K3S_VERSION="v1.20.4+k3s1" \
            INSTALL_K3S_EXEC="--tls-san {{.Values.node.hostname}}" \
            INSTALL_K3S_SELINUX_WARN="true" \
            sh -
       - curl https://raw.githubusercontent.com/helm/helm/master/scripts/get-helm-3 | bash
       - |
          helm -n fleet-system install --create-namespace --wait \
              fleet-crd https://github.com/rancher/fleet/releases/download/v0.3.3/fleet-crd-0.3.3.tgz
          helm -n fleet-system install --create-namespace --wait \
              fleet https://github.com/rancher/fleet/releases/download/v0.3.3/fleet-0.3.3.tgz
       # https://rancher.com/blog/2020/upgrade-k3s-kubernetes-cluster-system-upgrade-controller
       - kubectl apply -f https://raw.githubusercontent.com/rancher/system-upgrade-controller/v0.6.2/manifests/system-upgrade-controller.yaml
     - name: "Setup SSH keys"
       authorized_keys:
         root:
         - github:mudler
```

This will install k3s, fleet and suc.

2) Once the node is booted, and k3s started too, deploy the following plan:

```yaml
---
apiVersion: upgrade.cattle.io/v1
kind: Plan
metadata:
  name: cos-upgrade
  namespace: system-upgrade
  labels:
    k3s-upgrade: server
spec:
  concurrency: 1
  version: v1.19.4+k3s1
  nodeSelector:
    matchExpressions:
      - {key: k3s.io/hostname, operator: Exists}
  serviceAccountName: system-upgrade
  cordon: true
#  drain:
#    force: true
  upgrade:
    image: quay.io/costoolkit/test-images:os2
    command:
    - "/usr/sbin/suc-upgrade"
```

A pod should appear, if you attach to logs you will see it will upgrade the system and reboot

**Comment by davidcassany** [comment]
Cool, I have an equivalent step 1 running on the vanilla image already with an automatic deploy of cOS image and K3S, Fleet and SUC install at first boot after deployment. So applying this plan should do the trick :+1:

**Comment by mudler** [comment]
Ok, with fleet indeed was rather easy to trigger, that's what I've done manually to register a sample repo with the plan above:

```bash



cat > example.yaml << "EOF"
apiVersion: fleet.cattle.io/v1alpha1
kind: GitRepo
metadata:
  name: upgrade
  # This namespace is special and auto-wired to deploy to the local cluster
  namespace: fleet-local
spec:
  # Everything from this repo will be ran in this cluster. You trust me right?
  repo: "https://github.com/mudler/fleet-sample"
  branch: "main"
EOF

kubectl apply -f example.yaml

```

**Comment by davidcassany** [comment]
mm for some reason it fails trying to pull `quay.io/costoolkit/test-images:v1.19.4-k3s1` I would have expected `quay.io/costoolkit/test-images:os2` though, according to the manifest in https://github.com/mudler/fleet-sample

**Comment by mudler** [comment]
> mm for some reason it fails trying to pull `quay.io/costoolkit/test-images:v1.19.4-k3s1` I would have expected `quay.io/costoolkit/test-images:os2` though, according to the manifest in https://github.com/mudler/fleet-sample
> 
> Anyway, seeing that system-upgrade-controller gets into play using fleet is already a success :) Thanks @mudler 

:thinking:  might be the version of suc-controller, what are you running on?

**Comment by davidcassany** [comment]
> thinking might be the version of suc-controller, what are you running on?

Running on `cos/system` if this is what you are asking for. I just applied some user-data.yaml equivalent to what is in https://github.com/rancher-sandbox/cOS-toolkit/issues/339#issuecomment-872826814

**Comment by mudler** [comment]
> > thinking might be the version of suc-controller, what are you running on?
> 
> Running on `cos/system` if this is what you are asking for. I just applied some user-data.yaml equivalent to what is in [#339 (comment)](https://github.com/rancher-sandbox/cOS-toolkit/issues/339#issuecomment-872826814)

that's weird, did you deployed suc manually or by using the steps in the cloud-init config?  the cloud-init config provided in the comment installs a fixed version (`0.6.2`)

**Comment by davidcassany** [comment]
> that's weird, did you deployed suc manually or by using the steps in the cloud-init config? the cloud-init config provided in the comment installs a fixed version (`0.6.2`)

Oh I am installing a v0.4.0 i did the same snipped but using the link from the README, I did not check if it was up to date :facepalm:

**Comment by mudler** [comment]
I've created https://github.com/rancher-sandbox/cos-fleet-upgrades-sample as a full sample with steps. I'm going to update the docs in the cos-toolkit to mention it
